File: bytedance/25-reverse-nodes-in-k-group/25.go
package main

import ("fmt")

/**
 * Definition for singly-linked list.
 **/
type ListNode struct {
    Val int
    Next *ListNode
 }
 func reverseKGroup(head *ListNode, k int) *ListNode {
    hair := &ListNode{Next: head}
    pre := hair

    for head != nil {
        tail := pre
        for i := 0; i < k; i++ {
            tail = tail.Next
            if tail == nil {
                return hair.Next
            }
        }
        nx := tail.Next
        head, tail = rev(head, tail)
        tail.Next = nx
        pre.Next = head
        pre = tail
        head = tail.Next
    }
    return hair.Next
}

func rev(head *ListNode, tail *ListNode) (*ListNode, *ListNode) {
    prev := tail.Next
    p := head
    for prev != tail {
        nx := p.Next
        p.Next = prev
        prev = p
        p = nx   
    }
    return tail, head
}

func main() {
	node1 := &ListNode{Val: 1}
	node2 := &ListNode{Val: 2}
	node3 := &ListNode{Val: 3}
	node4 := &ListNode{Val: 4}
	node5 := &ListNode{Val: 5}
	node1.Next = node2
	node2.Next = node3
	node3.Next = node4
	node4.Next = node5
	res := reverseKGroup(node1, 2)
	for res != nil {
		fmt.Print(res.Val, "->")
		res = res.Next
	}
	fmt.Println()
}

File: bytedance/25-reverse-nodes-in-k-group/25.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseKGroup(self, head: Optional[ListNode],
                      k: int) -> Optional[ListNode]:
        hair = ListNode(0, head)
        pre = hair

        while head:
            tail = pre
            for i in range(k):
                tail = tail.next
                if not tail:
                    return hair.next

            nx = tail.next
            head, tail = self.rev(head, tail)
            pre.next = head
            tail.next = nx
            pre = tail
            head = tail.next
        return hair.next

    def rev(self, head: Optional[ListNode], tail: Optional[ListNode]):
        prev = tail.next
        p = head
        while prev != tail:
            nx = p.next
            p.next = prev
            prev = p
            p = nx
        return tail, head


if __name__ == '__main__':
    sol = Solution()